cajita/unit_test/tstSparseIndexSpace.hpp

```diff
@@ -0,0 +1,424 @@
+            {
+
+                EXPECT_EQ( test[i * size * size + j * size + k], true );
+            }
+}
+
+void testSparseIndexSpace()
+{
+    constexpr int dim_n = 3;
+    constexpr int size_per_dim = 16 * 4;
+
+    std::array<int, dim_n> size( { size_per_dim, size_per_dim, size_per_dim } );
+    int capacity = size_per_dim * size_per_dim;
+    float rehash_factor = 1.5;
+    SparseIndexSpace<TEST_EXECSPACE> sis( size, capacity, rehash_factor );
+
+    auto cbd = sis.cell_bits_per_tile_dim;
+    EXPECT_EQ( cbd, 2 );
+
+    auto cnd = sis.cell_num_per_tile_dim;
+    EXPECT_EQ( cnd, 4 );
+
+    auto cmd = sis.cell_mask_per_tile_dim;
+    EXPECT_EQ( cmd, 3 );
+
+    auto cbt = sis.cell_bits_per_tile;
+    EXPECT_EQ( cbt, 6 );
+
+    auto cnt = sis.cell_num_per_tile;
+    EXPECT_EQ( cnt, 64 );
+
+    Kokkos::View<int***, TEST_DEVICE> qid_res( "query_id", size[0], size[1],
+                                               size[2] );
+
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, size_per_dim ),
+        KOKKOS_LAMBDA( int i ) {
+            for ( int j = 0; j < size_per_dim; j++ )
+                for ( int k = 0; k < size_per_dim; k++ )
+                {
+                    sis.insertCell( i, j, k );
+                }
+        } );
+
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, size_per_dim ),
+        KOKKOS_LAMBDA( int i ) {
+            for ( int j = 0; j < size_per_dim; j++ )
+                for ( int k = 0; k < size_per_dim; k++ )
+                {
+                    auto qid = sis.queryCell( i, j, k );
+                    qid_res( i, j, k ) = qid;
+                }
+        } );
+
+    bool test[size_per_dim * size_per_dim * size_per_dim];
+    auto qid_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), qid_res );
+
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+            {
+                test[i * size[1] * size[2] + j * size[2] + k] = false;
+            }
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+            {
+                test[qid_mirror( i, j, k )] = true;
+            }
+    for ( int i = 0; i < size[0]; i++ )
+        for ( int j = 0; j < size[1]; j++ )
+            for ( int k = 0; k < size[2]; k++ )
+            {
+                EXPECT_EQ( test[i * size[2] * size[1] + j * size[2] + k],
+                           true );
+            }
+}
+
+//---------------------------------------------------------------------------//
+// RUN TESTS
+//---------------------------------------------------------------------------//
+TEST( sparse_grid, sparse_index_space_test )
+{
+    testAtomicOr();
+    testAtomicOrPro();
+    testTileSpace();
+    testBlockSpace<HashTypes::Naive>();
+    testBlockSpace<HashTypes::Morton>();
+    testSparseIndexSpace();
+}
```

Comment:
After the above changes with respect to memory management are made let's add another test which checks inserting, clearing, and inserting again. You could even do a small number of inserts, clear, reserve more capacity, insert again with more stuff, and make sure it is correct.

Reply:
Added